lib/hisq_paths_force_quda.cu

```diff
@@ -11,304 +11,545 @@ namespace quda {
 
   namespace fermion_force {
 
-    template <typename Arg> class FatLinkForce : public TunableKernel3D {
+    typedef std::reference_wrapper<GaugeField> GaugeField_ref;
+
+    struct dim_dir_pair {
+      int dim, dir;
+      static dim_dir_pair make_pair(int signed_dir) {
+        dim_dir_pair pr = { signed_dir > 3 ? 7 - signed_dir : signed_dir, signed_dir > 3 ? 0 : 1 };
+        return pr;
+      }
+
+      static dim_dir_pair invalid_pair() {
+        dim_dir_pair pr = { -1, -1 };
+        return pr;
+      }
+
+      int signed_dir() {
+        return is_invalid() ? -1 : ((dir == 1) ? dim : (7 - dim));
+      }
+
+      bool is_forward() const noexcept { return dir == 1; }
+      bool is_backward() const noexcept { return dir == 0; }
+      bool is_valid() const noexcept { return (is_forward() || is_backward()) && dim >= 0 && dim < 4; }
+      bool is_invalid() const noexcept { return !is_valid(); }
+    };
+
+    template <typename Arg> class OneLinkForce : public TunableKernel3D {
       Arg &arg;
-      const GaugeField &outA;
-      const GaugeField &outB;
-      const GaugeField &pMu;
-      const GaugeField &qMu;
+      const GaugeField &force;
+      const GaugeField &link;
+      unsigned int minThreads() const override { return arg.threads.x; }
+
+    public:
+      OneLinkForce(Arg &arg, const GaugeField &link, const GaugeField &force) :
+        TunableKernel3D(link, 2, 4),
+        arg(arg),
+        force(force),
+        link(link)
+      {
+        char aux2[16];
+        strcat(aux, comm_dim_partitioned_string());
+        strcat(aux, ",threads=");
```

Comment:
The `threads` number isn't needed in the `tuneKey` anymore, it can be nuked.  This was previously used to lift the degeneracy between say an 8^4 single-GPU local volume and a (4 + 2 * R)^4, with R = 2, multi-GPU local volume.  Since we now have `comm_dim_partitioned_string` in the `tuneKey` the `threads` label can be safely nuked.

Reply:
https://github.com/lattice/quda/pull/1367/commits/0f06ef4407130c0df053db25f74c7c7c3fe70e39